### ISSUE TITLE
Allow document selectors in static registration during initialize handshake.

### ISSUE THREAD

**Issue by kdvolder** [microsoft/language-server-protocol#751]
THe LSP protocol allows server to specify trigger characters for when the client should trigger auto-complete when user is typing. This is specified in 'CompletionOptions' part of the 'ServerCapabilities'. 

It is somewhat problematic for language servers that support multiple languages that you can only specify one set of trigger characters, which is then implicitly applicable for all the languages. 

The trigger characters tend to be language specific as they depend somewhat on the syntax of the language. Therefore the 'CompletionOptions' should allow a server to specify multiple sets of trigger characters and designate what language(s) each set applies to.

**Comment by rcjsuen** [comment]
Would it make sense to add something similar for `textDocument/onTypeFormatting`, `textDocument/signatureHelp`, and so on?

**Comment by kdvolder** [comment]
> Would it make sense to add something similar for textDocument/onTypeFormatting, textDocument/signatureHelp, and so on?

I would think so, but I only raised the issue about trigger characters because we directly experienced it as a problem. (We are not really affected by the onTypeFormatting and signatureHelp because our language server doesn't implement that part of the protocol).

**Comment by dbaeumer** [comment]
The protocol support this by registering completion support dynamically in the `initialized` notification send to the server. However the client needs to flag such a support using the `dynamicRegistration` property. The VS Code client does this for all capabilities. It is available for `onTypeFormatting` and `signatureHelp` as well.

@kdvolder will that work for you?

**Comment by dbaeumer** [comment]
See also: https://microsoft.github.io/language-server-protocol/specification#client_registerCapability and `CompletionRegistrationOptions` in this section: https://microsoft.github.io/language-server-protocol/specification#textDocument_completion

**Comment by Aerijo** [comment]
@dbaeumer I'm not sure how that would work; besides it being weird to have to swap trigger characters every time a different language file is used, how would the server know the user has swapped file type before the first request for that file? If the file was already open, there would be no opened notification.

**Comment by Gama11** [comment]
I think what @dbaeumer means is making multiple completion provider registrations with different language filters, no need to "swap" anything.

**Comment by Aerijo** [comment]
@Gama11 Oh right. Up until now I'd assumed 'dynamic registration' meant it could be turned off and on whenever.

@dbaeumer's suggestion makes sense now.

**Comment by Gama11** [comment]
It does still mean that, but it also allows making multiple registrations with different filters. :)

**Comment by kdvolder** [comment]
> @kdvolder will that work for you?

I beleave it will. I hadn't considered that route yet. It may be somewhat tenuous depending on how much all the different clients actually support the dynamic registration though. So for that reason it may be still worthwhile to consider a static alternative. (I beleave the complexity of managing registrations dynamically might well be a reason why many clients will just opt-out of implementing support for it).

**Comment by dbaeumer** [comment]
@kdvolder even if we support this in a static way it must be something a client must opt in. I tried to avoid this so far since I believe a client can equally opt into the dynamic registration.

**Comment by dbaeumer** [comment]
I changed the title to reflect the request.

**Comment by kdvolder** [comment]
>  way it must be something a client must opt in. I tried to avoid this so far since I believe a client can equally opt into the dynamic registration.

Yes, I understand that. 

Fortunately for my particular problem I was able to use dynamic registration. Turns out I only needed to set language-specific trigger characters for vscode. The Eclipse client, which is the other major client I cared about, does not support dynamic registration, however, it triggers completions correctly even if I just set a 'blanket' trigger on all supported languages at once.

There is probably a bug somewher there in how the vscode client, in how it combines triggers from multiple language servers (I should/will file a bug report for that if I can formulate it clearly). However, long story short, luckily for this very specific use-case, which is working around a bug in vscode completion triggers, the fact that only vscode supported the dynamic registration was good enough.

Anyhow... considering the static extension would be also 'opt-in' I', not so convinced now how hard I actually want this.

From server implementor's point of view. We now have two cases:

- client support dynamic completion registration
- client does not support dynamic registration

I have to detect this and provide some suitable way to handle this.

Adding a third static option will make this only harder in some sense. So perhaps it would be better to just put some pressure on client implementors to implement dynamic registration.

If you want to close this ticket, I'm fine with it.
